### PR TITLE
test(test): Adding fuzzer for ContainerPolicy

### PR DESCRIPTION
**Pull request by prady0t**
**Purpose of PR?**:

This is a fuzzer for `ContainerPolicy` https://github.com/kubearmor/KubeArmor/blob/a5f584c38ee7cd674d1e95a3f7ce7842212dfe12/KubeArmor/policy/policy.go#L24

For seed corpus I referred to [this](https://docs.kubearmor.io/kubearmor/documentation/security_policy_examples#:~:text=Block%20a%20specific%20executable) policy and removed `Metadata.Name` from it. 
With this we are testing if invalid policy results in `PolicyStatus_Invalid` or not. 
To run the fuzz test, 

```
cd KubeArmor/policy
go test -fuzz=Fuzz
```
Addresses #1367 

**Does this PR introduce a breaking change?**
NO

### CODE REVIEW

**Comment by prady0t** [comment]
Already found an issue with this. I ran the fuzzer for a while and It failed for this input : 
```json
{
    "0000": "00000000000000",
    "oBjeCt": {
        "metAdAtA": {
            "nAme": "0"
        }
    }
}
```
Which is totally understandable. To fix this we can also have a condition in fuzzer to check for status as `Applied` if we have `policyEvent.Object.Metadata.Name != ""` . Here's the log : 

```
--- FAIL: FuzzContainerPolicy (0.00s)
    --- FAIL: FuzzContainerPolicy/69798181ccc6962d (0.00s)
        ContainerPoilcy_fuzz_test.go:58: Unexpected status: Applied, [123 32 32 32 32 34 48 48 48 48 34 58 32 34 48 48 48 48 48 48 48 48 48 48 48 48 48 48 34 44 32 32 32 32 34 111 66 106 101 67 116 34 58 32 123 32 32 32 32 32 34 109 101 116 65 100 65 116 65 34 58 32 123 32 32 32 32 32 32 34 110 65 109 101 34 58 32 34 48 34 125 125 125]
FAIL
exit status 1
FAIL    github.com/kubearmor/KubeArmor/KubeArmor/policy 0.011s
```

Marking this as a draft for now.

**Comment by prady0t** [comment]
I've added a check for `pb.PolicyStatus_Applied` as well for JSON inputs with `Object.Metadata.Name != ""`. With this, our fuzzer should run until it finds an edge case where input results in `Unexpected Status`. We could have also added a condition to check if we get a `pb.PolicyStatus_Applied` status when `Object.Metadata.Name != ""` inside the fuzzer itself, instead of just the AND condition. But that would require to unmarshal the `data` and would be kind of meta (testing json.Unmarshal with json.Unmarshal).

If you think it's testing something useful, we can go ahead and merge it. Once we have this inside KubeArmor codebase, we can try writing config files so that we can integrate it with oss-fuzz. Running it locally won't give much fruitful results. 
In the [cilium fuzz testing report](https://docs.cilium.io/en/stable/_downloads/1bca8c3d8bb912f689017e7092afe682/CiliumFuzzingAudit2022.pdf) (mentioned in the upstream issue comments), we can see fuzzers run for 1000s of hours : 

Screenshot 2024-07-26 at 10 59 30 PM

Running our fuzzers with oss fuzz will help us improve it even better. We may also find some crucial bugs. This could be a starting point for how we can add more fuzzer. 

@daemon1024 @DelusionalOptimist Do let me know what you guys think of this.

File: KubeArmor/policy/ContainerPoilcy_fuzz_test.go
Comment (nyrahul) [inline KubeArmor/policy/ContainerPoilcy_fuzz_test.go:1]: Please fix the filename spelling .. Poilcy => Policy

File: KubeArmor/policy/ContainerPoilcy_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package policy
```
Comment (nyrahul) [inline KubeArmor/policy/ContainerPoilcy_fuzz_test.go]: Please add the SPDX header needed in all code files.

**Review by nyrahul — CHANGES_REQUESTED** [review]
Thank you for the PR

File: KubeArmor/policy/ContainerPoilcy_fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"context"
+	"testing"
+
+	pb "github.com/kubearmor/KubeArmor/protobuf"
+
+	tp "github.com/kubearmor/KubeArmor/KubeArmor/types"
```
Comment (nyrahul) [inline KubeArmor/policy/ContainerPoilcy_fuzz_test.go]:
```suggestion
	pb "github.com/kubearmor/KubeArmor/protobuf"
	tp "github.com/kubearmor/KubeArmor/KubeArmor/types"
```

**Comment by nyrahul** [comment]
I think this PR and the resultant image could be uses as input for OSS-Fuzz. How long did you test it on your local system? @prady0t

**Review by nyrahul — CHANGES_REQUESTED** [review]

**Comment by prady0t** [comment]
> I think this PR and the resultant image could be uses as input for OSS-Fuzz. How long did you test it on your local system? @prady0t

For about an hour. Do you want me to run it overnight? It might help us find any early bug or any problems with the fuzzer.

**Comment by nyrahul** [comment]
Running it in current state won't result in anything since the primary function is stubbed out.

**Comment by nyrahul** [comment]
The primary updates I made using the last commit is:

* Using the real `ParseAndUpdateContainerSecurityPolicy` to fuzz test. Removed the stubbed UpdateContainerPolicy.
* Fixing the fuzz test initial data

This is already showing up some issues during fuzz ... We need to check if those are genuine issues or need stubbing.

**Comment by prady0t** [comment]
Thanks a Lot for these changes! This makes much more sense now. Let me try to run these locally

**Comment by prady0t** [comment]
Sorry for the late reply. Fuzzer is failing for this input : 
```
[]byte("{\"oBjeCt\":{\"metAdAtA\":{\"nAme\":\"0\"},\"speC\":{\"seleCtor\":{\"mAtChLABels\":{\"\":\"\"}}}}}")

```
@nyrahul did you find similar results locally?

JSON format : 
```json
{
  "oBjeCt": {
    "metAdAtA": {
      "nAme": "0"
    },
    "speC": {
      "seleCtor": {
        "mAtChLABels": {
          "": ""
        }
      }
    }
  }
}

```

**Comment by prady0t** [comment]
The fuzzer is now running. Would like your reviews @nyrahul @daemon1024 @DelusionalOptimist

File: KubeArmor/core/containerPolicy_fuzz_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+	"testing"
+
+	"github.com/kubearmor/KubeArmor/KubeArmor/policy"
+	pb "github.com/kubearmor/KubeArmor/protobuf"
+)
+
+func FuzzContainerPolicy(f *testing.F) {
+	initialData := &pb.Policy{
+		Policy: []byte(`
```
Comment (daemon1024) [inline KubeArmor/core/containerPolicy_fuzz_test.go]: Let's use varied kinds of policy including file, network as seed values. https://github.com/kubearmor/KubeArmor/tree/main/tests/k8s_env/ksp/multiubuntu

File: KubeArmor/core/unorchestratedUpdates.go [hunk 4]
```diff
@@ -179,18 +179,18 @@ func (dm *KubeArmorDaemon) handlePolicyEvent(eventType string, createEndPoint bo
...
-
-		dm.Logger.UpdateSecurityPolicies("DELETED", newPoint)
-		dm.RuntimeEnforcer.UpdateSecurityPolicies(newPoint)
-
-		// delete endpoint if no containers or policies
+		if(endpointIdx != -1){
+			dm.EndPoints[endpointIdx] = newPoint
+		    dm.Logger.UpdateSecurityPolicies("DELETED", newPoint)
+		    dm.RuntimeEnforcer.UpdateSecurityPolicies(newPoint)
+					// delete endpoint if no containers or policies
```
Comment (daemon1024) [inline KubeArmor/core/unorchestratedUpdates.go]: We already found a bug and fixed it, Nice 🔥

**Review by daemon1024 — CHANGES_REQUESTED** [review]

**Comment by daemon1024** [comment]
@prady0t https://github.com/kubearmor/KubeArmor/actions/runs/11170637360/job/31168314822?pr=1797
gofmt fails ( even though the CI is passing ). Please run the formatter

File: KubeArmor/core/unorchestratedUpdates.go [hunk 5]
```diff
@@ -179,16 +179,16 @@ func (dm *KubeArmorDaemon) handlePolicyEvent(eventType string, createEndPoint bo
...
-
-		dm.Logger.UpdateSecurityPolicies("DELETED", newPoint)
-		dm.RuntimeEnforcer.UpdateSecurityPolicies(newPoint)
-
-		// delete endpoint if no containers or policies
-		if len(newPoint.Containers) == 0 && len(newPoint.SecurityPolicies) == 0 {
-			dm.EndPoints = append(dm.EndPoints[:endpointIdx], dm.EndPoints[endpointIdx+1:]...)
-			// since the length of endpoints slice reduced
-			endpointIdx--
+		if endpointIdx != -1 {
```
Comment (DelusionalOptimist) [inline KubeArmor/core/unorchestratedUpdates.go]:
```suggestion
		if endpointIdx >= 0 {
```

**Review by DelusionalOptimist — CHANGES_REQUESTED** [review]
@prady0t looks mostly good. One small fix.
Also, can you squash the commits?

**Pull request closed without merging**